Add support for QUAD-ZIG-SW from smarthjemmet.dk

mortenmoulder opened the pull request:
## Proposed change
Add support for smarthjemmet.dk's QUAD-ZIG-SW device.


## Additional information
Works with V1 and V2 firmware.

## Checklist

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works

mortenmoulder committed:
Add support for QUAD-ZIG-SW from smarthjemmet.dk

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

mortenmoulder committed:
Fix docstrings

mortenmoulder committed:
Merge branch 'dev' of https://github.com/mortenmoulder/zha-device-handlers into dev

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

mortenmoulder committed:
Update file name and add tests

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

mortenmoulder committed:
Fix docstring

mortenmoulder committed:
Merge branch 'dev' of https://github.com/mortenmoulder/zha-device-handlers into dev

mortenmoulder commented:
All should be fixed now. Sorry for all the commits.

mortenmoulder commented:
@TheJulianJES I see that a new release was created yesterday, and mine was not included. Any idea why? The only checks failing are code coverage, which shouldn't prevent it from being merged. Was really hoping to get this quirk implemented in the next release, as I'm guessing I have to wait a few more weeks now.

TheJulianJES commented:
There's a few things on this PR that I'd likely change. I just haven't found the time to properly review this PR and comment all small change suggestions. I still plan to do that, of course, but there are also other PRs.
Yesterday's release was just to get some bigger ZHA related changes into nightly HA builds. The quirks related changes were mostly ones that didn't need feedback.

TheJulianJES committed:
Remove empty `__init__`

TheJulianJES committed:
Remove unused `_current_state`

TheJulianJES committed:
Use walrus operator for `action`

TheJulianJES committed:
Refactor quirk `CustomDevice` classes

Follows the style of other quirks where there's proper quirk defined in a `CustomDevice` class that's inherited from in other quirks that just change the `signature`, but keep the `replacement` and device triggers.

TheJulianJES committed:
Replace with v2 quirk

TheJulianJES commented:
Ok, so I've pushed multiple commits now. The "refactor" commit changes the structure of the (so-called v1) quirk to have a  `v1` device class with `signature`, `replacement`, and device triggers.
Then, a class for firmware version 2 is added that inherits from the first device class. It just changes the `signature`.
This is similar to the structure we have in other quirks. 

The last commit (as of now) "Replace with v2 quirk" completely replaces the `CustomDevice` v1 quirks with a v2 quirk.
For it, we don't need to have a fixed `signature` and `replacement`. `.replaces(CR2032PowerConfigurationCluster)` will automatically replace all clusters with the same cluster id on endpoint 1. It's used to replace the default `PowerConfiguration` cluster with the custom `CR2032PowerConfigurationCluster`.

`.replace_cluster_occurrences(CustomMultistateInputCluster)` will look at all endpoints and replace all clusters with the same cluster id (of `MultistateInput`) it can find.

Please test if that v2 quirk works as expected.

---

With the v1 quirk before, the `MultistateInput` cluster was always an output/client cluster in the original signature. In the `replacement`, it was a input/server cluster then. It was also missing on endpoint 1 in the replacement. I don't think that was intentional?

If the replacement `CustomMultistateInputCluster` needs to be an input/server cluster (so the role swapped compared the original signature), we can do it like this in a v2 quirk (but for all endpoints):
```python
    .removes(MultistateInput.cluster_id, cluster_type=ClusterType.Client, endpoint_id=1)  # remove output cluster
    .adds(CustomMultistateInputCluster, endpoint_id=1)  # default cluster_type is server, so it'll add an input cluster
```

TheJulianJES committed:
Remove test

The test here is no longer really needed for the v2 quirk. There's no real point in testing if all the correct clusters are included.
Ideally, we would send messages from the device and see if they are converted to "ZHA events" as expected in a test.

TheJulianJES commented on zhaquirks/smarthjemmet/quadzigsw.py:
```diff
@@ -0,0 +1,94 @@
+class CustomMultistateInputCluster(CustomCluster, MultistateInput):
+    """Multistate input cluster."""
+
+    async def bind(self):
+        """Prevent bind."""
+        return zigpy.zcl.foundation.Status.SUCCESS
+
+    async def unbind(self):
+        """Prevent unbind."""
+        return zigpy.zcl.foundation.Status.SUCCESS
+
+    async def _configure_reporting(self, *args, **kwargs):
+        """Prevent remote configure reporting."""
+        return (
+            zigpy.zcl.foundation.ConfigureReportingResponse.deserialize(b"\x00")[0],
+        )
```
Unrelated to converting the whole quirk to a v2 quirk:
Are these necessary? We already set `SKIP_CONFIGURATION = True`.

mortenmoulder commented:
@TheJulianJES Hi and thanks for your effort. Unfortunately it's not working.

When I am talking about V1 and V2 in terms of firmware, it's because I have one physical device with two different firmware versions. The signature of the devices are completely different, which is why I need to be able to support both signatures on the same device. V1 of the firmware came out back in 2023, and now I've updated it in late 2024. Does that make sense?

For all I care, Quirk V1 or Quirk V2 doesn't matter to me. Quirk V2 looks better with the QuirkBuilder - definitely, but unfortunately it's not working as intended in this case.

Regards to your "It was also missing on endpoint 1 in the replacement. I don't think that was intentional" it actually was. The physical device has 5 buttons, but only 4 of them should work as MultistateInput. The last is used for resetting the device, which is done on a firmware level.

mortenmoulder commented:
And for good measure, here's the old firmware (firmware V1 from 2023)'s device signature:

```
{
  "node_descriptor": {
    "logical_type": 2,
    "complex_descriptor_available": 0,
    "user_descriptor_available": 0,
    "reserved": 0,
    "aps_flags": 0,
    "frequency_band": 8,
    "mac_capability_flags": 128,
    "manufacturer_code": 4447,
    "maximum_buffer_size": 80,
    "maximum_incoming_transfer_size": 160,
    "server_mask": 0,
    "maximum_outgoing_transfer_size": 160,
    "descriptor_capability_field": 0
  },
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0000",
        "0x0001",
        "0x0007"
      ],
      "output_clusters": [
        "0x0000",
        "0x0001",
        "0x0012"
      ]
    },
    "2": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "3": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "4": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "5": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    }
  },
  "manufacturer": "smarthjemmet.dk",
  "model": "QUAD-ZIG-SW",
  "class": "zigpy.device.Device"
}
```

and here is the new firmware V2 (from late 2024) device signature:

```
{
  "node_descriptor": {
    "logical_type": 2,
    "complex_descriptor_available": 0,
    "user_descriptor_available": 0,
    "reserved": 0,
    "aps_flags": 0,
    "frequency_band": 8,
    "mac_capability_flags": 128,
    "manufacturer_code": 4447,
    "maximum_buffer_size": 80,
    "maximum_incoming_transfer_size": 160,
    "server_mask": 0,
    "maximum_outgoing_transfer_size": 160,
    "descriptor_capability_field": 0
  },
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0000",
        "0x0001"
      ],
      "output_clusters": [
        "0x0000"
      ]
    },
    "2": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007",
        "0x0012"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "3": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007",
        "0x0012"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "4": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007",
        "0x0012"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "5": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007",
        "0x0012"
      ],
      "output_clusters": [
        "0x0012"
      ]
    }
  },
  "manufacturer": "smarthjemmet.dk",
  "model": "QUAD-ZIG-SW",
  "class": "QUAD-ZIG-SW.QUAD_ZIG_SW"
}
```

mortenmoulder commented:
@TheJulianJES Help is not really wanted. What I originally had worked perfectly. Only thing not passing is the code coverage. Is that a requirement to get it merged?

TheJulianJES commented:
"Help wanted" just means this needs to be looked again. A v2 quirk should work.
And yes, the code should be covered fully. That's not really a big issue though.

mortenmoulder commented:
Okay I see. Is it for you to look at "at some point" as a reminder? Or someone else? My question really is, do I need to take action?

TheJulianJES commented:
I'll take another look again at this soon. In the mean time, you can try to see if you can make the v2 quirk work if you want.
E.g. by copying/pasting the lines at the bottom from [this comment](https://github.com/zigpy/zha-device-handlers/pull/3400#issuecomment-2420918826) into the quirk (replacing the `replace_cluster_occurrences`).
(Adding those two lines for endoints (1?) 2, 3, 4, 5.)

> Regards to your "It was also missing on endpoint 1 in the replacement. I don't think that was intentional" it actually was. The physical device has 5 buttons, but only 4 of them should work as MultistateInput.

`range(1, 6)` gives us device triggers for five buttons though. That was also in the [original code](https://github.com/zigpy/zha-device-handlers/blob/df836e275a64b6d719635e7a86a42c194ee5971e/zhaquirks/smarthjemmet/quadzigsw.py).
Isn't that one batch of device triggers too many then?
It's also a bit weird, because the original device signature has the MultistateInputCluster on all 5 endpoints (as an output cluster). Could be firmware thing of course, still weird though..

> Unfortunately it's not working.

On which HA version did you test the v2 quirk? Did you test it by replacing the whole quirks library or just inserting the one quirks file as a "custom quirk" (using the ZHA config option) (recommended for testing this).
Also, did the quirk apply at least? E.g., in device settings in HA or in the device signature, was there a `CustomDeviceV2` quirk used?

mortenmoulder commented:
@TheJulianJES I am doing the custom quirk with the ZHA config option. It was matched to the signature of one of the devices I tested (the old V1 firmware), but no events were shown when I pressed the buttons.

Also, I appreciate the help. Let me know if there is anything I can do.

OptionHenrik commented:
> @TheJulianJES I am doing the custom quirk with the ZHA config option. It was matched to the signature of one of the devices I tested (the old V1 firmware), but no events were shown when I pressed the buttons.
> 
> Also, I appreciate the help. Let me know if there is anything I can do.

@mortenmoulder … was any of your quirk versions working when you added them as custom quirks? I think we are many users waiting for ZHA support to happen and I would like to try deploying a quirk file if your can point to a version you tested to be working.

@TheJulianJES what is missing yet before the PR can be accepted? Hope you will have the time to look at this again. Really appreciate the work you are doing.

mortenmoulder commented:
@OptionHenrik Yes, the quirk was working perfectly before @TheJulianJES's changes. Feel free to go look at the commit just before his changes, and grab the file from there to give it a go yourself.

TheJulianJES committed:
Remove copied code from `LocalDataCluster`

We already "skip configuration" during pairing, so we don't need this.

TheJulianJES committed:
Adjust device trigger range to range(1, 5)

The device only has 5 endpoints. Endpoint 2 to 5 have MultistateInputClusters. This will account for that, as `ENDPOINT_ID: i + 1` with i ranging from 1 to 4.

TheJulianJES committed:
Remove output clusters, re-add as input clusters

This isn't really correct, but it's what the original quirk did.

TheJulianJES committed:
Use `AttributeDefs` in `_update_attribute`

TheJulianJES commented:
I've made the changes suggested in this comment: https://github.com/zigpy/zha-device-handlers/pull/3400#issuecomment-2420918826
So, I've replicated what the original quirk did (removes actual `MultistateInput` output clusters and inserts our virtual `MultistateInput` clusters as input clusters. This technically isn't correct, but let's see if that's the issue with the v2 quirk.)

The original PR also created a device trigger for endpoint 6, even though there the device only goes up to endpoint 5.

Please try and see if these changes work now.

TheJulianJES commented on zhaquirks/smarthjemmet/quadzigsw.py:
```diff
@@ -0,0 +1,87 @@
+                event_args = {VALUE: value}
+                self.listener_event(ZHA_SEND_EVENT, action, event_args)
+
+            super()._update_attribute(0, action)
```
If the changes work, please also try and see what these additional changes to this line do:
- replacing 0 with `MultistateInput.AttributeDefs.present_value.id` in this line (sensor entity in HA?)
- completely removing this line

I believe this was copied from other quirks where it doesn't actually work, as there is no attribute with `id=0` on this cluster.

OptionHenrik commented:
> @OptionHenrik Yes, the quirk was working perfectly before @TheJulianJES's changes. Feel free to go look at the commit just before his changes, and grab the file from there to give it a go yourself.

Works like a charm … great work so far even though it’s only a custom quirk. I hope you guys get it integrated into the standard ZHA soon

TheJulianJES commented:
Please also test the v2 quirk now. It was updated a few days ago.

radhus commented:
@mortenmoulder can confirm your quirk works with just a model name change for MULTI-ZIG-SW as well! Great job.

I'll try to find time to test out @TheJulianJES changes later.

bangert commented:
Any news here on the current version in this PR?

OptionHenrik commented:
> Any news here on the current version in this PR?

See my previous comment. Works like a charm with ZHA if you install it as a quirk.

TheJulianJES commented:
It would also be ideal to know if the updated v2 quirk version works correctly now. Then, we'd be able to finally merge this (after I add some tests).

bangert commented:
@mortenmoulder  would this quirk also work for MULTI-ZIG-SW?

bangert commented:
trying to get the latest version of this quirk to work with MULTI-ZIG-SW it doesnt work, and i get the following in the HA logs:
```
2024-12-13 01:04:27.370 ERROR (MainThread) [homeassistant.components.logbook.processor] Error with zha describe event for zha_event
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/logbook/processor.py", line 259, in _humanify
    data = describe_event(event_cache_get(row))
  File "/usr/src/homeassistant/homeassistant/components/zha/logbook.py", line 73, in async_describe_zha_event
    event_type = event_type.replace("_", " ").title()
                 ^^^^^^^^^^^^^^^^^^
AttributeError: 'bool' object has no attribute 'replace'
```

this is on HA 2024.12.2

bangert commented:
okay - some more feedback. the current quirk (latest version on this PR) does not actually for me on the sister device MULTI-ZIG-SW, but it is recognised if i add two lines. The first is line 28
```
MODEL_ALT_ID = "MULTI-ZIG-SW"
```
and its usage on line 58/59
```
.applies_to(MANUFACTURER, MODEL_ALT_ID)
```

meaning it will show up as "MULTI-ZIG-SW by smarthjemmet.dk" and show `Quirk: zigpy.quirks.v2.CustomDeviceV2` in zigbee info.

Nothing happens when pushing any of the buttons though.

bangert commented:
with this patch the quadzigsw.py quirk works for me on a MULTI_ZIG_SW:
```patch
--- quadzigsw-orig.py   2025-02-01 06:26:31.211611131 +0100
+++ quadzigsw.py        2025-02-01 06:27:26.903118134 +0100
@@ -25,6 +25,7 @@
 
 MANUFACTURER = "smarthjemmet.dk"
 MODEL_ID = "QUAD-ZIG-SW"
+MODEL_ALT_ID = "MULTI-ZIG-SW"
 
 ACTION_TYPE = {
     0: COMMAND_RELEASE,
@@ -43,20 +44,19 @@
 
 
 class CustomMultistateInputCluster(CustomCluster, MultistateInput):
-    """Multistate input cluster."""
+    """CustomMultistate input cluster."""
 
     def _update_attribute(self, attrid, value):
         super()._update_attribute(attrid, value)
         if attrid == MultistateInput.AttributeDefs.present_value.id:
-            if action := ACTION_TYPE.get(value) is not None:
+            if (action := ACTION_TYPE.get(value)) is not None:
                 event_args = {VALUE: value}
                 self.listener_event(ZHA_SEND_EVENT, action, event_args)
 
-            super()._update_attribute(0, action)
-
 
 (
     QuirkBuilder(MANUFACTURER, MODEL_ID)
+    .applies_to(MANUFACTURER, MODEL_ALT_ID)
     .skip_configuration()
     .replaces(CR2032PowerConfigurationCluster)
     .removes(MultistateInput.cluster_id, cluster_type=ClusterType.Client, endpoint_id=2)
```

TheJulianJES commented:
@bangert Thanks! The issue with the walrus operator was an oversight on my part. I'll merge these changes and add tests for this device soon, then we can finally merge this quirk.

TheJulianJES committed:
Merge branch 'dev' into mortenmoulder-dev

TheJulianJES committed:
Fix walrus (thanks @bangert)

TheJulianJES committed:
Remove updating attribute `0` on press

TheJulianJES committed:
Also apply to `MULTI-ZIG-SW` model

TheJulianJES committed:
Add tests

TheJulianJES committed:
Combine if statements

bangert commented:
@TheJulianJES i have re-tested the latest version on a MULTI-ZIG-SW v1 firmware: it still works.

@mortenmoulder @OptionHenrik could you test with QUAD-ZIG-SW and/or v2 firmware?

TheJulianJES reviewed (approved):
Thanks!

mortenmoulder commented:
@TheJulianJES Thanks for merging. I have verified that MULTI-ZIG-SW works, however, it's not possible to change the cluster "on the fly", like I am able to using Zigbee2MQTT: 

![image](https://github.com/user-attachments/assets/740dd01f-70b6-4b28-bcb2-7c740d644e9a)

Do you think that's possible? Here is the MULTI-ZIG-SW code for Zigbee2MQTT: https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/custom_devices_diy.ts#L736-L759

On another note, QUAD-ZIG-SW does not work. I'm not sure why. It gets paired and shows up correctly, but I cannot see any events. Here is the signature for V1 of the firmware:

```
{
  "node_descriptor": {
    "logical_type": 2,
    "complex_descriptor_available": 0,
    "user_descriptor_available": 0,
    "reserved": 0,
    "aps_flags": 0,
    "frequency_band": 8,
    "mac_capability_flags": 128,
    "manufacturer_code": 64447,
    "maximum_buffer_size": 80,
    "maximum_incoming_transfer_size": 160,
    "server_mask": 0,
    "maximum_outgoing_transfer_size": 160,
    "descriptor_capability_field": 0
  },
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0000",
        "0x0001",
        "0x0007"
      ],
      "output_clusters": [
        "0x0000",
        "0x0001",
        "0x0012"
      ]
    },
    "2": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "3": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "4": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    },
    "5": {
      "profile_id": "0x0104",
      "device_type": "0xfffe",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0012"
      ]
    }
  },
  "manufacturer": "smarthjemmet.dk",
  "model": "QUAD-ZIG-SW",
  "class": "zigpy.quirks.v2.CustomDeviceV2"
}
```

and here is the signature for the V2 of the firmware:

```
{
  "node_descriptor": {
    "logical_type": 2,
    "complex_descriptor_available": 0,
    "user_descriptor_available": 0,
    "reserved": 0,
    "aps_flags": 0,
    "frequency_band": 8,
    "mac_capability_flags": 128,
    "manufacturer_code": 64447,
    "maximum_buffer_size": 80,
    "maximum_incoming_transfer_size": 160,
    "server_mask": 0,
    "maximum_outgoing_transfer_size": 160,
    "descriptor_capability_field": 0
  },
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0000",
        "0x0001",
        "0x0007"
      ],
      "output_clusters": [
        "0x0000",
        "0x0001",
        "0x0006",
        "0x0012"
      ]
    },
    "2": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0006",
        "0x0012"
      ]
    },
    "3": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0006",
        "0x0012"
      ]
    },
    "4": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0006",
        "0x0012"
      ]
    },
    "5": {
      "profile_id": "0x0104",
      "device_type": "0x0006",
      "input_clusters": [
        "0x0007"
      ],
      "output_clusters": [
        "0x0006",
        "0x0012"
      ]
    }
  },
  "manufacturer": "smarthjemmet.dk",
  "model": "QUAD-ZIG-SW",
  "class": "zigpy.quirks.v2.CustomDeviceV2"
}
```

Only difference is the device_type going from 0xfffe to 0x0006.

The device has 5 buttons, but one of them is used for resetting the device. I see that when I create an automation, I have 5 buttons. 

Neither V1 or V2 works. Let me know if you want to make some changes, then I will give it a go.
